### PR TITLE
Add a Rust implementation of `coz.h`

### PR DESCRIPTION
**Pull request by alexcrichton**
After seeing a recording of Emery's recent talk at Strange Loop I was
super impressed with `coz` and immediately wanted to try it out with
Rust code! I was curious to see if `coz` could help us out with
profiling either Cargo, Rust's build system, or perhaps the compiler
itself, `rustc`. In any case I ended up creating
https://github.com/alexcrichton/coz-rs as a repository to house an
implementation of `coz.h` for Rust so `coz` could be used to easily work
with Rust executables.

I toyed around a bit and Emery [mentioned] that this would be good
support to have upstream once it's working, and I think it's working
now! I haven't been able to get a ton of useful results myself yet, but
I'm still figuring out the best way to ask `coz` questions and the best
way to instrument programs myself.

Nonetheless I think that the Rust support here is correct in the sense
that it's a faithful translation of `coz.h`, and now it's just a problem
with me learning how best to instrument programs.

This commit adds a new `rust` directory with a crate (Rust parlance for
"library") which provides macros that allow using `coz` instrumentation
similar as you would in C, except in idiomatic Rust style. I've
published the crate on Rust's package registry, crates.io, as
[`coz`][crate], and I don't mind keeping it there or transferring it to
y'all, whichever you'd prefer! I also don't mind being on the hook for
reviewing any Rust-related changes, or if y'all need to jettison the
Rust support in the future I don't mind that either and can give it a
home!

[mentioned]: https://github.com/alexcrichton/coz-rs/issues/2
[crate]: https://crates.io/crates/coz

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    ($name:expr) => {{
+        static COUNTER: $crate::Counter = $crate::Counter::end($name);
+        COUNTER.increment();
+    }};
+}
+
+/// Perform one-time per-thread initialization for `coz`.
+///
+/// This may not be necessary to call, but for good measure it's recommended to
+/// call once per thread in your application near where the thred starts.
```
Comment (bjorn3) [inline rust/src/lib.rs]: *thread

**Comment by ccurtsinger** [comment]
It looks like the rust check is failing on TravisCI because `cargo` was not installed. Could you update the TravisCI config to install that?

File: .travis.yml [hunk 2]
```diff
@@ -27,3 +27,12 @@ install:
 # Run the make check target (add tests later)
 script:
   - make check USE_SYSTEM_COZ=1
+
+# Test the Rust support as well
+matrix:
+  include:
+    - rust: stable
```
Comment (bjorn3) [inline .travis.yml]:
```suggestion
    - language: rust
      rust: stable
```

**Comment by alexcrichton** [comment]
Sure thing, done!

**Comment by bjorn3** [comment]
Travis CI is happy :tada:

**Comment by alexcrichton** [comment]
Thanks @emeryberger!

Please feel free to cc me on any Rust related issues/changes, I don't mind doing my share of the maintenance for this :)

**Pull request merged**